### PR TITLE
ENH: Add `fs=` parameter to filter design functions

### PR DESCRIPTION
**Pull request by endolith**
This is just a more convenient way of specifying normalized frequencies for digital filters.  Instead of having to remember that `butter` uses `fc/(fs/2)`, while `freqz` uses `fc/(fs/(2*pi))`, for instance, you can just specify the frequencies in Hz and the sampling rate in Hz.

Old way:

```
>>> fc = 1000
>>> fs = 3000
>>> b, a = butter(2, fc/(fs/2))
>>> abs(freqz(b, a, [fc/(fs/(2*pi))])[1])
array([ 0.70710678])
```

New way:

```
>>> fc = 1000
>>> fs = 3000
>>> b, a = butter(2, fc, fs=fs)
>>> abs(freqz(b, a, [fc], fs=fs)[1])
array([ 0.70710678])
```

I proposed this a while ago and no one seemed interested, but it's easy enough to implement so we'll see what people think.

Examples of confusion:

- https://dsp.stackexchange.com/q/46915/29
- https://dsp.stackexchange.com/q/24989/29#comment45751_24992
- https://mail.python.org/pipermail/scipy-user/2008-March/015944.html

### CODE REVIEW

**Comment by WarrenWeckesser** [comment]
Thanks for this, it is a much needed enhancement.  I'll try to do a review sometime this weekend (nag me if I don't!).
 
One thing I'll note now is that unit tests are needed for the new option.

**Comment by larsoner** [comment]
Agreed, even if it's redundant the new code is much more readable. +1 from me

**Comment by larsoner** [comment]
I've optimistically marked this for 1.1 (mid-March)

**Comment by endolith** [comment]
@WarrenWeckesser 

> One thing I'll note now is that unit tests are needed for the new option.

Yeah I wanted to see if people liked it before doing the rest of the work.  Sounds like people do, though.

**Comment by endolith** [comment]
Instead of default of `fs=None`, I could make it `fs=2` or `fs=`2*pi` or whatever is appropriate for each function?  I don't really have an opinion on which is better.

Also can be added to
- iirnotch, iirpeak
- buttord, cheb1ord, etc.
- group_delay
- iirdesign
- band_stop_obj ?

**Comment by larsoner** [comment]
I prefer explicit `fs=2.` and similar but IIRC @WarrenWeckesser prefers `fs=None` as an alias because it can be easier to wrap.

Maybe we can both be happy by using `fs=2.` as the default but support `fs=None` as "use the default" (in the code this is the one-liner `fs = 2. if fs is None else fs`)? That way introspection of the call immediately shows the default (rather than having to look in the docstring; what I like) and functions that wrap these can still pass `None` to get default behavior (what I think @WarrenWeckesser wants)?

As for what needs to be changed, I'd recommend taking a brief scan of the `test_*` functions in `scipy.signal` and seeing where else `fs=` would make things simpler. But I also don't mind multiple PRs if we find other cases in the future, so I don't think this *needs* to be exhaustive.

**Comment by endolith** [comment]
Should freqz, ellipord, etc. return w in fs units if fs specified?

**Comment by endolith** [comment]
Added examples that use fs and sos to filter a real signal:

https://5541-1460385-gh.circle-artifacts.com/0/home/ubuntu/scipy/doc/build/html-scipyorg/generated/scipy.signal.butter.html

Also added test for some functions but not all yet

**Comment by endolith** [comment]
So doctest failed because of 

```
Failed example:
    ax1.axis([0, 1, -2, 2])
Expected nothing
Got:
    [0, 1, -2, 2]
```

But there are other examples with 

```
>>> ax.axis((10, 1000, -100, 10))
```

that don't cause failures?

**Comment by larsoner** [comment]
Looks like `ax.axis` is explicitly ignored:

https://github.com/scipy/scipy/blob/master/tools/refguide_check.py#L518

This could be improved/extended with e.g. `regex` that permits names like `ax1` and `ax2`, or just use `# doctest:+SKIP` or similar with the offending lines.

**Comment by larsoner** [comment]
Maybe the thing to do is to change it to `.axis(`from `ax.axis`. It should be general enough. Same with `plt.plot(`, it seems like this should be `.plot(`.

This should probably be a separate PR, though, so that it gets proper discussion as it modifies the test suite.

**Comment by ev-br** [comment]
Please no doctest:+SKIP, just fix the refguide-checker.

**Comment by WarrenWeckesser** [comment]
I can live with `fs=2.0` (or whatever the appropriate constant is for the function being changed).

**Comment by WarrenWeckesser** [comment]
@endolith asked

> Should freqz, ellipord, etc. return w in fs units if fs specified?

My expectation was that `fs` would set the units for everything--input and output--so the frequencies returned would be in the same units as `fs`.  I'm trying to think of a reason for it to be otherwise, but so far I'm not succeeding.  Can anyone make a compelling argument for the returned frequencies to not match `fs`?

**Comment by larsoner** [comment]
> Can anyone make a compelling argument for the returned frequencies to not match fs?

As long as the new defaults give the same old results, we should be good to go on this. I have the same expectation as you, that all outputs would be modified appropriately.

**Comment by WarrenWeckesser** [comment]
The functions `periodogram`, `welch`, `csd`, `coherence`, `spectrogram` and `stft` all accept the argument `fs` already, and the frequencies that they return are in the same units as `fs`.  Given that existing behavior, for API consistency the functions updated in this pull request should also return frequencies in the same units as `fs`.

**Comment by endolith** [comment]
I'm not sure what to do about the failures.

`operator.index(array([123]))` is raising `DeprecationWarning: converting an array with ndim > 0 to an index will result in an error in the future`

but it's *supposed* to produce an error, and does in Python 3.6.3, but not in Python 3.4.6 the test is running in?

**Comment by larsoner** [comment]
> but it's supposed to produce an error, and does in Python 3.6.3, but not in Python 3.4.6 the test is running in?

Different Travis runs use different NumPy versions, this is probably to blame. The 3.4.6 appears to use NumPy 1.8.2. The code for `freqz` probably needs to be improved to deal with this case.

**Comment by endolith** [comment]
So this goes back to https://github.com/scipy/scipy/pull/7351 and how to verify int parameters.

I made a chart of how different types are handled here: https://stackoverflow.com/a/48940855/125507 

I did this using `.index()` since that seems to be what people want.

File: scipy/signal/filter_design.py [hunk 1]
```diff
@@ -162,9 +162,13 @@ def freqs(b, a, worN=None, plot=None):
...
-    elif isinstance(worN, int):
-        N = worN
-        w = findfreqs(b, a, N)
+    elif np.ndim(worN) == 0:
+        try:
+            N = operator.index(worN)
+        except TypeError:
+            pass
+        else:
+            w = findfreqs(b, a, N)
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: You could save on branching/lines a little bit with `else` instead of `elif` here:
```
else:
    try:
        N = operator.index(worN)
    except TypeError:
        w = worN
    else:
        w = findfreqs(b, a, N)
```

File: scipy/signal/filter_design.py [hunk 2]
```diff
@@ -299,12 +304,15 @@ def freqz(b, a=1, worN=None, whole=False, plot=None):
         A callable that takes two arguments. If given, the return parameters
         `w` and `h` are passed to plot. Useful for plotting the frequency
         response inside `freqz`.
+    fs : float
+        The sampling frequency of the digital system.
+        .. versionadded:: 1.1.0
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: Need an extra newline before the `..` directive otherwise it doesn't render properly (not seen as separate paragraph)

https://5706-1460385-gh.circle-artifacts.com/0/home/ubuntu/scipy/doc/build/html-scipyorg/generated/scipy.signal.freqz.html#scipy.signal.freqz

File: scipy/signal/filter_design.py [hunk 3]
```diff
@@ -299,12 +304,15 @@ def freqz(b, a=1, worN=None, whole=False, plot=None):
...
 
     Returns
     -------
     w : ndarray
-        The normalized frequencies at which `h` was computed, in
-        radians/sample.
+        The frequencies at which `h` was computed.  If `fs` is specified,
+        these are in the same units.  Otherwise they are in radians/sample.
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: This is a case where setting `fs=2*pi` makes things simpler. Here you can just say they are in units of the sample rate, and you can get rid of the conditionals below. We can make `fs=None` behave like `2 * pi` if this would make achieving "use-the-default" behavior easier for people. @WarrenWeckesser WDYT?

**Review by larsoner — COMMENTED** [review]
`operator.index` is indeed what we settled on for `int` checks.

I made some small comments on the code, let's see what @WarrenWeckesser says about the `fs` default, and once that's settled I'll take a look at the examples and tests.

File: scipy/signal/filter_design.py (hunk 1 above)
Comment (endolith) [reply scipy/signal/filter_design.py]: It needs to do the ndim check to interpret `array([8])` as "measure at point 8", rather than "measure at 8 points" since apparently older numpy versions returned .index() on 1-element N-D arrays.

File: scipy/signal/filter_design.py (hunk 3 above)
Comment (endolith) [reply scipy/signal/filter_design.py]: The `worN=None` arguments follow the same pattern, right?  They actually default to numbers, too.

Comment (larsoner) [reply scipy/signal/filter_design.py]: Yes, I am arguing against that design choice as it hurts argument introspection. I think we should ideally change this everywhere, so at least avoid propagating the problem here.

Comment (larsoner) [reply scipy/signal/filter_design.py]: I just remembered this was discussed in #7788. I'll try to move that forward so we can proceed here.

Comment (endolith) [reply scipy/signal/filter_design.py]: Maybe that should be done in a separate PR then?  (And changing `worN=512` as well?)

**Comment by endolith** [comment]
Documentation should be good now.  Changed some examples to use new format 
and left others alone.  As usual, the documentation and tests make up 95% of the work.  :grin:

Please double-check that I didn't copy and paste the wrong parameter names, 1 vs pi, etc.  I think it's all correct, but I did a lot of copying and pasting.

So only thing left is to decide on default parameters?  (Or postpone that for another PR)

**Comment by larsoner** [comment]
Yes for consistency let's do `None` for now and we can change them all later if need be.

File: scipy/signal/filter_design.py (hunk 1 above)
Comment (WarrenWeckesser) [reply scipy/signal/filter_design.py]: Using, for example, `worN=3.0`, results in `UnboundLocalError: local variable 'w' referenced before assignment`

File: scipy/signal/filter_design.py (hunk 3 above)
Comment (WarrenWeckesser) [reply scipy/signal/filter_design.py]: Sorry for disappearing from this conversation.  If no one else objects to `fs=2*np.pi`, then go ahead.  (Of course, introspection of the signature will then show `fs=6.283185307179586`.)

File: scipy/signal/filter_design.py (hunk 1 above)
Comment (endolith) [reply scipy/signal/filter_design.py]: I modified it to fix the unbound error. Now it calls index twice, but that's not a big deal? Otherwise I'd have to duplicate other things and it would be harder to read. Still need to write tests for this condition

Comment (larsoner) [reply scipy/signal/filter_design.py]: Calling `index` twice seems fine if it leads to the simplest code.

File: scipy/signal/filter_design.py (hunk 3 above)
Comment (WarrenWeckesser) [reply scipy/signal/filter_design.py]: That's a neat idea, but I don't think the extra code is worth it just to fix a cosmetic issue.

**Comment by endolith** [comment]
Since https://github.com/scipy/scipy/pull/8584 was merged, my to-do list is:

- [ ] Finish writing tests for every w/N/fs/fft configuration
- [ ] Change `fs` defaults to explicit numbers and remove `if None` code.

**Comment by larsoner** [comment]
I think the plan is to leave the `if None` wrappers for backward compat (at least). Might as well keep them in docstrings, too.

**Comment by endolith** [comment]
(Also shouldn't it raise ValueError if whole is specified at the same time as an array_like w?  Currently whole is just ignored)

**Comment by larsoner** [comment]
I would just add a note to the `whole` docstring that it is ignored if `worN` is an array.

**Comment by endolith** [comment]
Changing defaults, and noticed that for all the filter design functions, setting both `analog=True` and `fs=2` currently raises ValueError, so if we're changing to default `fs=2`, fs will be ignored?  This seems like an argument for `fs=None` for those functions, actually.

**Comment by larsoner** [comment]
Yes in #7788 we converged on the idea that if a default value depends on some other input values, using `None` and explaining this behavior in the docstring is good.

**Comment by endolith** [comment]
Ok I think I'm done

File: scipy/signal/filter_design.py [hunk 4]
```diff
@@ -39,6 +39,17 @@ class BadCoefficients(UserWarning):
 abs = absolute
 
 
+def _is_int_type(x):
+    if np.ndim(x) != 0:
+        return False
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: Why is this necessary? I think we have been using just `operator.index` as below everywhere else. And `np.array([1]).__index__()` gives a TypeError.

File: scipy/signal/filter_design.py (hunk 3 above)
Comment (larsoner) [reply scipy/signal/filter_design.py]: Back to this line of code -- we should just say that the frequencies are in units of `fs` (and if you want, add that the default `fs=2*pi` implies the units are radians/sample)

File: scipy/signal/filter_design.py [hunk 5]
```diff
@@ -471,19 +493,26 @@ def freqz_zpk(z, p, k, worN=512, whole=False):
...
+
+        .. versionadded:: 1.1.0
 
     Returns
     -------
     w : ndarray
-        The normalized frequencies at which `h` was computed, in
-        radians/sample.
+        The frequencies at which `h` was computed.  If `fs` is specified,
+        these are in the same units.  Otherwise they are in radians/sample.
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: Same here (and other replicates), they should always be in units of `fs`.

File: scipy/signal/filter_design.py [hunk 6]
```diff
@@ -1917,20 +1982,38 @@ def iirfilter(N, Wn, rp=None, rs=None, btype='band', analog=False,
...
+    >>> ax.set_title('Chebyshev Type II bandpass frequency response')
+    >>> ax.set_xlabel('Frequency [Hz]')
+    >>> ax.set_ylabel('Amplitude [dB]')
+    >>> ax.axis((10, 1000, -100, 10))
+    >>> ax.grid(which='both', axis='both')
+    >>> plt.show()
+
+    Create a digital filter with the same properties, in a system with
+    sampling rate of 2000 Hz, and plot the frequency response.  (Second-order
+    sections implementation is required for a filter of this order):
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: ... required to ensure stability of a filter ...

File: scipy/signal/filter_design.py [hunk 7]
```diff
@@ -4150,17 +4429,18 @@ def iirpeak(w0, Q):
     >>> plt.show()
     """
 
-    return _design_notch_peak_filter(w0, Q, "peak")
+    return _design_notch_peak_filter(w0, Q, "peak", fs)
 
 
-def _design_notch_peak_filter(w0, Q, ftype):
+def _design_notch_peak_filter(w0, Q, ftype, fs=2):
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: 2.

File: scipy/signal/filter_design.py [hunk 8]
```diff
@@ -3995,7 +4270,7 @@ def besselap(N, norm='phase'):
     return asarray([]), asarray(p, dtype=complex), float(k)
 
 
-def iirnotch(w0, Q):
+def iirnotch(w0, Q, fs=2):
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: 2. (explicit float)

File: scipy/signal/filter_design.py [hunk 9]
```diff
@@ -2452,12 +2548,33 @@ def butter(N, Wn, btype='low', analog=False, output='ba'):
     >>> plt.axvline(100, color='green') # cutoff frequency
     >>> plt.show()
 
+    Generate a signal made up of 10 Hz and 20 Hz, sampled at 1 kHz
+
+    >>> t = np.linspace(0, 1, 1000)  # 1 second
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: You need `endpoint=False` otherwise the effective sample rate is actually 999 Hz. This common mistake is one reason I usually opt for explicit division of sample numbers by `fs`, i.e. `np.arange(1000) / 1000.`

File: scipy/signal/filter_design.py [hunk 10]
```diff
@@ -2538,12 +2663,33 @@ def cheby1(N, rp, Wn, btype='low', analog=False, output='ba'):
     >>> plt.axhline(-5, color='green') # rp
     >>> plt.show()
 
+    Generate a signal made up of 10 Hz and 20 Hz, sampled at 1 kHz
+
+    >>> t = np.linspace(0, 1, 1000)  # 1 second
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: `endpoint=False` here, too, or use division

File: scipy/signal/filter_design.py [hunk 11]
```diff
@@ -2619,12 +2773,33 @@ def cheby2(N, rs, Wn, btype='low', analog=False, output='ba'):
     >>> plt.axhline(-40, color='green') # rs
     >>> plt.show()
 
+    Generate a signal made up of 10 Hz and 20 Hz, sampled at 1 kHz
+
+    >>> t = np.linspace(0, 1, 1000)  # 1 second
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: endpoint

File: scipy/signal/filter_design.py [hunk 12]
```diff
@@ -2712,12 +2895,34 @@ def ellip(N, rp, rs, Wn, btype='low', analog=False, output='ba'):
     >>> plt.axhline(-5, color='green') # rp
     >>> plt.show()
 
+    Generate a signal made up of 10 Hz and 20 Hz, sampled at 1 kHz
+
+    >>> t = np.linspace(0, 1, 1000)  # 1 second
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: endpoint

File: scipy/signal/filter_design.py [hunk 13]
```diff
@@ -499,48 +528,57 @@ def freqz_zpk(z, p, k, worN=512, whole=False):
...
+    >>> ax1.grid()
 
     >>> ax2 = ax1.twinx()
     >>> angles = np.unwrap(np.angle(h))
-    >>> plt.plot(w, angles, 'g')
-    >>> plt.ylabel('Angle (radians)', color='g')
-    >>> plt.grid()
+    >>> ax2.plot(w, angles, 'g')
+    >>> ax2.set_ylabel('Angle [radians]', color='g')
+
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: [Looks good](https://6409-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.signal.freqs_zpk.html#scipy.signal.freqs_zpk)

File: scipy/signal/filter_design.py [hunk 14]
```diff
@@ -1917,20 +1982,38 @@ def iirfilter(N, Wn, rp=None, rs=None, btype='band', analog=False,
...
+    >>> sos = signal.iirfilter(17, [50, 200], rs=60, btype='band',
+    ...                        analog=False, ftype='cheby2', fs=2000,
+    ...                        output='sos')
+    >>> w, h = signal.sosfreqz(sos, 2000, fs=2000)
+    >>> fig = plt.figure()
+    >>> ax = fig.add_subplot(1, 1, 1)
     >>> ax.semilogx(w, 20 * np.log10(abs(h)))
     >>> ax.set_title('Chebyshev Type II bandpass frequency response')
-    >>> ax.set_xlabel('Frequency [radians / second]')
+    >>> ax.set_xlabel('Frequency [Hz]')
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: [Looks good](https://6409-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.signal.iirfilter.html#scipy.signal.iirfilter)

File: scipy/signal/filter_design.py [hunk 15]
```diff
@@ -2452,12 +2548,33 @@ def butter(N, Wn, btype='low', analog=False, output='ba'):
...
+    format when filtering, to avoid numerical error with transfer function
+    (``ba``) format):
+
+    >>> sos = signal.butter(10, 15, 'hp', fs=1000, output='sos')
+    >>> filtered = signal.sosfilt(sos, sig)
+    >>> ax2.plot(t, filtered)
+    >>> ax2.set_title('After 15 Hz high-pass filter')
+    >>> ax2.axis([0, 1, -2, 2])
+    >>> ax2.set_xlabel('Time [seconds]')
+    >>> plt.show()
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: [Looks good](https://6409-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.signal.butter.html#scipy.signal.butter) but could probably benefit from a `plt.tight_layout()` to avoid axes overlap

File: scipy/signal/filter_design.py [hunk 16]
```diff
@@ -2538,12 +2663,33 @@ def cheby1(N, rp, Wn, btype='low', analog=False, output='ba'):
...
+    format when filtering, to avoid numerical error with transfer function
+    (``ba``) format):
+
+    >>> sos = signal.cheby1(10, 1, 15, 'hp', fs=1000, output='sos')
+    >>> filtered = signal.sosfilt(sos, sig)
+    >>> ax2.plot(t, filtered)
+    >>> ax2.set_title('After 15 Hz high-pass filter')
+    >>> ax2.axis([0, 1, -2, 2])
+    >>> ax2.set_xlabel('Time [seconds]')
+    >>> plt.show()
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: [Same thing about tight_layout](https://6409-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.signal.cheby1.html#scipy.signal.cheby1)

File: scipy/signal/filter_design.py [hunk 17]
```diff
@@ -2619,12 +2773,33 @@ def cheby2(N, rs, Wn, btype='low', analog=False, output='ba'):
...
+    format when filtering, to avoid numerical error with transfer function
+    (``ba``) format):
+
+    >>> sos = signal.cheby2(12, 20, 17, 'hp', fs=1000, output='sos')
+    >>> filtered = signal.sosfilt(sos, sig)
+    >>> ax2.plot(t, filtered)
+    >>> ax2.set_title('After 17 Hz high-pass filter')
+    >>> ax2.axis([0, 1, -2, 2])
+    >>> ax2.set_xlabel('Time [seconds]')
+    >>> plt.show()
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: [Same](https://6409-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.signal.cheby2.html#scipy.signal.cheby2)

File: scipy/signal/filter_design.py [hunk 18]
```diff
@@ -2712,12 +2895,34 @@ def ellip(N, rp, rs, Wn, btype='low', analog=False, output='ba'):
...
+    format when filtering, to avoid numerical error with transfer function
+    (``ba``) format):
+
+    >>> sos = signal.ellip(8, 1, 100, 17, 'hp', fs=1000, output='sos')
+    >>> filtered = signal.sosfilt(sos, sig)
+    >>> ax2.plot(t, filtered)
+    >>> ax2.set_title('After 17 Hz high-pass filter')
+    >>> ax2.axis([0, 1, -2, 2])
+    >>> ax2.set_xlabel('Time [seconds]')
+    >>> plt.show()
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: [One more](https://6409-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.signal.ellip.html#scipy.signal.ellip)

File: scipy/signal/filter_design.py [hunk 19]
```diff
@@ -4124,14 +4406,11 @@ def iirpeak(w0, Q):
...
-    >>> w0 = f0/(fs/2)  # Normalized Frequency
     >>> # Design peak filter
-    >>> b, a = signal.iirpeak(w0, Q)
+    >>> b, a = signal.iirpeak(f0, Q, fs)
 
     >>> # Frequency response
-    >>> w, h = signal.freqz(b, a)
-    >>> # Generate frequency axis
-    >>> freq = w*fs/(2*np.pi)
+    >>> freq, h = signal.freqz(b, a, fs=fs)
```
Comment (larsoner) [inline scipy/signal/filter_design.py]: Nice simplifications, [looks good](https://6409-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.signal.iirpeak.html#scipy.signal.iirpeak).

**Review by larsoner — COMMENTED** [review]
Other than some minor tweaks, LGTM. No red in tests indicates good backward compatibility.

File: scipy/signal/filter_design.py (hunk 4 above)
Comment (endolith) [reply scipy/signal/filter_design.py]: It doesn't give a type error in older numpy versions if I remember correctly

Comment (endolith) [reply scipy/signal/filter_design.py]: Should add a comment explaining that though

**Comment by endolith** [comment]
@larsoner  Ok, I did those all as separate commits, I can squash things or whatever I best.

**Comment by larsoner** [comment]
Maintainers can squash when merging, so I don't think you need to.

**Comment by endolith** [comment]
Example in  https://6409-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.signal.cheb1ord.html could be modified to use fs format for readability.

Not affected by this PR, but https://6409-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.signal.sosfreqz.html#scipy.signal.sosfreqz graphs look bad, should be the same axes and grid lines

**Comment by larsoner** [comment]
Agreed, feel free to fix them here if you want

File: scipy/signal/filter_design.py [hunk 20]
```diff
@@ -1917,20 +1988,39 @@ def iirfilter(N, Wn, rp=None, rs=None, btype='band', analog=False,
...
-    response:
+    Generate a 17th-order Chebyshev II analog bandpass filter from 50 Hz to
+    200 Hz and plot the frequency response:
 
     >>> from scipy import signal
     >>> import matplotlib.pyplot as plt
 
-    >>> b, a = signal.iirfilter(17, [50, 200], rs=60, btype='band',
-    ...                         analog=True, ftype='cheby2')
+    >>> b, a = signal.iirfilter(17, [2*np.pi*50, 2*np.pi*200], rs=60,
```
Comment (WarrenWeckesser) [inline scipy/signal/filter_design.py]: Scaling these frequencies by 2*pi is not necessary.  This example of an analog filter was fine the way it was.

Comment (WarrenWeckesser) [reply scipy/signal/filter_design.py]: OK, my comment stands.  There is no fixed absolute scale for an analog filter, so no need to convert whatever the input units are to rad/s.

If you prefer to interpret the values in the example as Hz, the only change needed in the example is to change the label of the `x` axis to `Frequency [Hz]`.

Comment (endolith) [reply scipy/signal/filter_design.py]: Yeah I changed it so it's equivalent to the following digital example

Comment (WarrenWeckesser) [reply scipy/signal/filter_design.py]: OK, to be consistent about using Hz, for the analog example, you can leave `Wn` as `[50, 200]` and just change the label of the `x` axis as I mentioned above.  There is no need for converting back and forth to rad/s; that will just confuse readers.

Comment (endolith) [reply scipy/signal/filter_design.py]: > There is no fixed absolute scale for an analog filter

@WarrenWeckesser But the b, a coefficients represent a transfer function that's based on angular frequency, not ordinary frequency, right?  It's not just the input frequencies and output plot that matter for the example, it's the intermediate values too

Comment (WarrenWeckesser) [reply scipy/signal/filter_design.py]: With an analog design, if you change the units of the input frequencies, the coefficients change, but only to account for the change in scale.  The coefficients are not based on any specific units.  For example,

```
In [80]: b, a = signal.iirfilter(5, [50, 200], rs=60, btype='band', analog=True, ftype='cheby2')

In [81]: a
Out[81]: 
array([1.00000000e+00, 2.08556501e+02, 7.17476259e+04, 9.75002759e+06,
       1.70949761e+09, 1.54504252e+11, 1.70949761e+13, 9.75002759e+14,
       7.17476259e+16, 2.08556501e+18, 1.00000000e+20])
```
Multiply both input frequencies by 2*pi and design the filter again:

```
In [82]: b2, a2 = signal.iirfilter(5, [2*np.pi*50, 2*np.pi*200], rs=60, btype='band', analog=True, ftype='cheby2')

In [83]: a2
Out[83]: 
array([1.00000000e+00, 1.31039914e+03, 2.83248274e+06, 2.41849643e+09,
       2.66432973e+12, 1.51300296e+15, 1.05183522e+18, 3.76933662e+20,
       1.74279571e+23, 3.18304331e+25, 9.58956006e+27])
```
`a2` looks different, but that's because the coefficients include powers of the scale factor 2*pi.  That scale can be taken out to recover `a`:
``` 
In [84]: a2/(2*np.pi)**np.arange(0, len(a))
Out[84]: 
array([1.00000000e+00, 2.08556501e+02, 7.17476259e+04, 9.75002759e+06,
       1.70949761e+09, 1.54504252e+11, 1.70949761e+13, 9.75002759e+14,
       7.17476259e+16, 2.08556501e+18, 1.00000000e+20])
```
So unlike the digital case, which has a sample period that determines a scale, there is no natural scale for the analog case.  As long as you are consistent, everything will just work.  If you specify the design frequencies using, say, gradians/fortnight, and then evaluate the frequency response at a set of frequencies expressed in gradians/fortnight, it will work.

File: scipy/signal/tests/test_filter_design.py [hunk 21]
```diff
@@ -1189,6 +1400,24 @@ def test_analog(self):
...
+    def test_fs_param(self):
+        wp = [4410, 11025]
+        ws = [2205, 13230]
+        rp = 3
+        rs = 80
+        fs = 44100
+        N, Wn = buttord(wp, ws, rp, rs, False, fs=fs)
+        b, a = butter(N, Wn, 'bandpass', False, fs=fs)
+        w, h = freqz(b, a, fs=fs)
+        assert_array_less(-rp - 0.1,
```
Comment (WarrenWeckesser) [inline scipy/signal/tests/test_filter_design.py]: Do you need these tolerances (0.1) to be so large?  When I run this code, I get

```
In [81]: dB(h[np.logical_and(wp[0] <= w, w <= wp[1])]).min()
Out[81]: -3.0000001585928446

In [82]: dB(h[np.logical_or(w <= ws[0], ws[1] <= w)]).max()
Out[82]: -83.080125250056454
```
so the passband check would pass with a much smaller tolerance, and the stop band passes without a tolerance.

File: scipy/signal/tests/test_filter_design.py [hunk 22]
```diff
@@ -1189,6 +1400,24 @@ def test_analog(self):
...
+                          dB(h[np.logical_and(wp[0] <= w, w <= wp[1])]))
+        assert_array_less(dB(h[np.logical_or(w <= ws[0], ws[1] <= w)]),
+                          -rs + 0.1)
+
+        assert_equal(N, 18)
+        assert_allclose(Wn, [4409.722701715714, 11025.47178084662],
```
Comment (WarrenWeckesser) [inline scipy/signal/tests/test_filter_design.py]: How did you derive such precise values for the expected results?  Add a comment explaining where these numbers came from, so future maintainers can reproduce those expected values.

**Comment by larsoner** [comment]
@endolith I'm going to remove the 1.1.0 tag as it will branch Thursday, and it looks like there are a few kinks to work out still.

File: scipy/signal/tests/test_filter_design.py (hunk 22 above)
Comment (endolith) [reply scipy/signal/tests/test_filter_design.py]: I think the hardcoded values were to detect changes in the behavior of whether the filter "hugs" the inner or outer boundary.  

`buttord`'s hugging behavior was changed in https://github.com/scipy/scipy/pull/3235#issuecomment-33882609 

Discussion was here: https://github.com/scipy/scipy/issues/3219#issuecomment-32824200 

some discussion of 0.1 dB tolerances in both of those

Comment (endolith) [reply scipy/signal/tests/test_filter_design.py]: Anyway can those test changes be a separate PR, since they're about the `...ord` functions and not about the fs parameter?

File: scipy/signal/filter_design.py (hunk 20 above)
Comment (endolith) [reply scipy/signal/filter_design.py]: But the transfer function coefficients are based on the Laplace transform and the definition *s = σ + jω*, which is angular frequency.  

For example, `iirfilter(1, 1, btype='low', analog=True)` produces `(array([1.]), array([1., 1.]))` which represents the transfer function H(s) = 1/(s+1), which has a −3 dB point at 1 rad/unit.

I've never seen anyone define *s* using ordinary frequency, and I don't think we should do that just to make the example look simpler.  The docstring examples should demonstrate how `iirfilter` is normally used, which is angular frequency, and introducing scale factors will just create confusion.

Comment (WarrenWeckesser) [reply scipy/signal/filter_design.py]: @endolith, OK, my extended brain fart is over.  You're right, the continuous time functions should have frequencies expressed in rad/[time unit].  Thanks for your patience!

Comment (ilayn) [reply scipy/signal/filter_design.py]: > I've never seen anyone define s using ordinary frequency, 

Not that I want to argue against but we actually started exactly doing that because rad/s is just a scalar too much and means nothing in continuous time. First thing we all do, in matlab, for example is to switch to Hz. because that's what we are measuring and seeing on the plot. Same argument holds for dB. The extra 20 factor means nothing to almost(!) anyone.

Anyways, I guess we should stick to the conventions so I would also choose rad/s here. 

The main reason for my comment is (😃) that we started to see the GitHub unicorn. Do you mind maybe squashing some of the earlier commits such that the page is a bit less dense?

Comment (ilayn) [reply scipy/signal/filter_design.py]: I don't even know if that would help but just to keep the page still loadable worth a try I guess.

**Comment by endolith** [comment]
So any other issues?  It is finished from my end.  Changes to `*ord` testing can be another PR, since that requires discussion and pre-dates this PR anyway

**Comment by endolith** [comment]
*poke*  

Do you need anything else from me?

**Comment by larsoner** [comment]
@WarrenWeckesser do you have time to look again to see if your concerns are addressed?

**Comment by WarrenWeckesser** [comment]
@endolith sorry for not getting back to this sooner.  I'll take another look.  @larsoner and any other folks working on `signal`: if I don't get back to this by this Saturday, don't wait for me.  If you think it looks good, go ahead and merge.

**Review by WarrenWeckesser — APPROVED** [review]
OK, looks good.

**Comment by WarrenWeckesser** [comment]
I forgot to comment on one change that is needed: the `.. versionadded:: 1.1.0` directives in the docstrings of the functions that have a new `fs` parameter need to be updated to `1.2.0`.

**Comment by ilayn** [comment]
I guess everybody's happy now. In it goes then. Thanks a lot @endolith !!
